test/acceptance/user.controller.acceptance.ts

```diff
@@ -0,0 +1,120 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/example-shopping
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {createClientForHandler, supertest, expect} from '@loopback/testlab';
+import {RestServer} from '@loopback/rest';
+import {ShoppingApplication} from '../..';
+
+describe('UserController', () => {
+  let app: ShoppingApplication;
+  let server: RestServer;
+  let client: supertest.SuperTest<supertest.Test>;
+  let userId: string;
+  const user = {
+    email: 'test@loopback.io',
+    password: 'p4ssw0rd',
+    firstname: 'Example',
+    surname: 'User',
+  };
+
+  before(givenAnApplication);
+
+  before(givenARestServer);
+
+  before(async () => {
+    await app.boot();
+    await app.start();
+  });
+
+  before(() => {
+    client = createClientForHandler(server.requestHandler);
+  });
+
+  after(async () => {
+    await app.stop();
+  });
+
+  it('creates new user when POST /users is invoked', async () => {
+    const res = await client
+      .post('/users')
+      .send(user)
+      .expect(200);
+
+    // Assertions
+    expect(res.body.email).to.equal('test@loopback.io');
+    expect(res.body.firstname).to.equal('Example');
+    expect(res.body.surname).to.equal('User');
+    expect(res.body).to.have.property('id');
+    expect(res.body).to.not.have.property('password');
+    userId = res.body.id;
+  });
+
+  it('throws error for POST /users with a missing email', async () => {
+    const res = await client
+      .post('/users')
+      .send({
+        password: 'p4ssw0rd',
+        firstname: 'Example',
+        surname: 'User',
+      })
+      .expect(422);
+
+    const errorText = JSON.parse(res.error.text);
+    expect(errorText.details[0].info.missingProperty).to.equal('email');
+  });
+
+  it('throws error for POST /users with an invalid email', async () => {
+    const res = await client
+      .post('/users')
+      .send({
+        email: 'test@loop&back.io',
+        password: 'p4ssw0rd',
+        firstname: 'Example',
+        surname: 'User',
+      })
+      .expect(422);
+
+    expect(res.body.message).to.equal('invalid email');
+  });
+
+  it('throws error for POST /users with a missing password', async () => {
+    const res = await client
+      .post('/users')
+      .send({
+        email: 'test@loopback.io',
+        firstname: 'Example',
+        surname: 'User',
+      })
+      .expect(422);
+
+    const errorText = JSON.parse(res.error.text);
+    expect(errorText.details[0].info.missingProperty).to.equal('password');
+  });
+
+  it('throws error for POST /users with a string', async () => {
+    await client
+      .post('/users')
+      .send('hello')
+      .expect(415);
+  });
+
+  it('returns a user with given id when POST /user/{id} is invoked', async () => {
+    const userWithId = Object.assign({}, user, {id: userId});
```

Comment:
Instead of setting `userId` in a separate test, I think it's better for the user to be created in the test (through the repository), set the `userId`, and then call the endpoint. This allows tests to not be dependent upon each other.

Also, the created user instances need to be deleted after each tests